rust/processor/src/processors/marketplaces_processor.rs

```diff
@@ -0,0 +1,1048 @@
+                    // so we need to parse collection offer metadata from the event
+                    let collection_offer_filled_metadata = CollectionOfferEventMetadata {
+                        collection_offer_id: collection_offer.collection_offer_id.clone(),
+                        fee_schedule_id: collection_offer.fee_schedule_id.clone(),
+                        buyer: collection_offer.buyer.clone(),
+                        item_price: collection_offer.item_price.clone(),
+                        collection_metadata: collection_metadata.unwrap(),
+                    };
+                    collection_offer_filled_metadatas.insert(
+                        collection_offer_filled_metadata.collection_offer_id.clone(),
+                        collection_offer_filled_metadata,
+                    );
+                }
+            }
+
+            let mut object_metadatas: HashMap<String, ObjectCore> = HashMap::new();
+            let mut listing_metadatas: HashMap<String, ListingMetadata> = HashMap::new();
+            let mut fixed_price_listings: HashMap<String, FixedPriceListing> = HashMap::new();
+            let mut listing_token_v1_containers: HashMap<String, TokenType> =
+                HashMap::new(); // String to ListingTokenV1Container
+            let mut token_offer_metadatas: HashMap<String, TokenOfferMetadata> = HashMap::new();
+            let mut token_offer_v1s: HashMap<String, TokenOfferV1> = HashMap::new();
+            let mut token_offer_v2s: HashMap<String, TokenOfferV2> = HashMap::new();
+            let mut collection_offer_metadatas: HashMap<String, CollectionOfferMetadata> =
+                HashMap::new();
+            let mut collection_offer_v1s: HashMap<String, CollectionOfferV1> = HashMap::new();
+            let mut collection_offer_v2s: HashMap<String, CollectionOfferV2> = HashMap::new();
+            let mut auction_listings: HashMap<String, AuctionListing> = HashMap::new();
+
+            // Loop 2
+            // Parse out all the listing, auction, bid, and offer data from write set changes.
+            // This is a bit more complicated than the other parsers because the data is spread out across multiple write set changes,
+            // so we need a first loop to get all the data.
+            for (wsc_index, wsc) in txn.info.as_ref().unwrap().changes.iter().enumerate() {
+                match wsc.change.as_ref().unwrap() {
+                    Change::WriteResource(write_resource) => {
+                        let move_resource_address = standardize_address(&write_resource.address);
+                        let move_resource_type = write_resource.type_str.clone();
+                        let move_resource_type_address =
+                            write_resource.r#type.as_ref().unwrap().address.clone();
+                        if !move_resource_type_address.eq(marketplace_contract_address) {
+                            continue;
+                        }
```

Comment:
We can also wrap this in a helper function. [Example](https://github.com/aptos-labs/aptos-indexer-processors/blob/b2325e585d68c5e2089527504943445906b4711d/rust/processor/src/models/token_v2_models/v2_token_utils.rs#L357)